### PR TITLE
S3FeedStorage supports custom endpoint of object storage.

### LINKED ISSUE #4871: S3 FeedStorage does not support custom S3 like storages
### Description
Storage options for S3 feed storage is not exporting to custom S3 like storage as supported by image pipeline. The endpoint `botocore` trying to hit is AWS's servers and it is not respecting the endpoint set using `AWS_ENDPOINT_URL` and `AWS_REGION_NAME`

### Steps to Reproduce

1. set all the necessary parameters in `settings.py` file such as  `AWS_ACCESS_KEY_ID,AWS_SECRET_ACCESS_KEY, AWS_ENDPOINT_URL, AWS_REGION_NAME, FEED_STORAGE_S3_ACL`.

```
AWS_ACCESS_KEY_ID = 'XXXXXXXXXXXX'
AWS_SECRET_ACCESS_KEY = 'XXXXXXXXXXXXXXX'
AWS_USE_SSL = True  # or True (None by default)
AWS_VERIFY = True  # or True (None by default)
AWS_ENDPOINT_URL = 'https://example.abc.com'  <-- custom s3 like cloud
AWS_REGION_NAME = 'us-east' 
FEED_STORAGE_S3_ACL = 'private'
```
2.  Then setup feeds using
```
FEEDS = {
    's3://bucket/path/to/file-%(name)s-%(time)s.json': {
        'format': 'json',
        'encoding': 'utf8',
        'store_empty': False,
        'fields': None,
        'indent': 4,
        'item_export_kwargs': {
            'export_empty_fields': True,
        }
    }
}
```
3. ``` scrapy crawl spider```

**Expected behavior:** 
This should be able to store files to other S3 like cloud, as it does with `ImagePipeline`

**Actual behavior:**
```
DEBUG: Starting new HTTPS connection (1): bucket.s3.amazonaws.com:443 <---- I think this is the problem

Response body:
b'<?xml version="1.0" encoding="UTF-8"?>\n<Error><Code>InvalidAccessKeyId</Code><Message>The AWS Access Key Id you provided does not exist in our records.</Message><AWSAccessKeyId>XXXXXXXXXXXXXXXXXX</AWSAccessKeyId>

[scrapy.extensions.feedexport] ERROR: Error storing json feed (36 items) in: s3://bucket/path/to/file-spider-2020-11-04T11-19-05.json

botocore.exceptions.ClientError: An error occurred (InvalidAccessKeyId) when calling the PutObject operation: The AWS Access Key Id you provided does not exist in our records.
```

**Reproduces how often:** 
Always

### Versions
```
Scrapy       : 2.4.0
lxml         : 4.6.1.0
libxml2      : 2.9.10
cssselect    : 1.1.0
parsel       : 1.6.0
w3lib        : 1.22.0
Twisted      : 20.3.0
Python       : 3.7.7 (v3.7.7:d7c567b08f, Mar 10 2020, 02:56:16) - [Clang 6.0 (clang-600.0.57)]
pyOpenSSL    : 19.1.0 (OpenSSL 1.1.1h  22 Sep 2020)
cryptography : 3.2.1
Platform     : Darwin-19.6.0-x86_64-i386-64bit
```

### LINKED ISSUE #4997: Custom Endpoint for S3FeedStorage
## Summary

Scrapy documentation says that, scrapy supports other object storages via `AWS_ENDPOINT_URL`. Unfortunately, custom object storage endpoint does not support in feed exporter. I fixed the bug.

## Motivation

We cannot pass s3 endpoint url to s3 client. Because of that, we could not use digitalocean spaces. Now, it allows to pass endpoint url. So, we are able to use other object storages.

## Describe alternatives you've considered

We resolved the bug in our environment by creating custom concrete class from S3FeedStorage.

## Additional context

- [Scrapy Docs](https://docs.scrapy.org/en/latest/topics/settings.html?highlight=AWS_ENDPOINT_URL#aws-endpoint-url)
- #4871

### PR DESCRIPTION
**Pull request by bkayranci**
Scrapy documentation says that, scrapy supports other object storages via `AWS_ENDPOINT_URL`. Unfortunately, custom object storage endpoint does not support in feed exporter.

## Related Links
- Fixes #4997
- Fixes #4871 
- https://docs.scrapy.org/en/latest/topics/settings.html?highlight=AWS_ENDPOINT_URL#aws-endpoint-url

### CODE REVIEW

**Review by Gallaecio — APPROVED** [review]
Thanks!

File: docs/topics/feed-exports.rst [hunk 1]
```diff
@@ -192,9 +192,10 @@ passed through the following settings:
  * :setting:`AWS_ACCESS_KEY_ID`
  * :setting:`AWS_SECRET_ACCESS_KEY`
 
-You can also define a custom ACL for exported feeds using this setting:
+You can also define a custom ACL and custom endpoint for exported feeds using this setting:
 
  * :setting:`FEED_STORAGE_S3_ACL`
+ * :setting:`AWS_ENPOINT_URL`
```
Comment (wRAR) [inline docs/topics/feed-exports.rst:198]: Typo

Comment (bkayranci) [reply docs/topics/feed-exports.rst]: thanks @wRAR I fixed it.

**Comment by edgarrmondragon** [comment]
Hi @Gallaecio. Is something blocking the merge on this one? If so, perhaps I can address any issues in case @bkayranci is not available :smiley:

**Comment by Gallaecio** [comment]
It looks good to me, and @wRAR ’s feedback has been addressed, so I think we can merge.

**Comment by bkayranci** [comment]
> The documentation lacks usage info... do I need only AWS_ENDPOINT_URL set to my s3-like url in the `settings.py` file? I guess I'll need the aws secret and access key variables as well?

The documentation are available for [access](https://docs.scrapy.org/en/latest/topics/settings.html?highlight=AWS_ACCESS_KEY_ID#aws-access-key-id) and [secret](https://docs.scrapy.org/en/latest/topics/settings.html?highlight=AWS_SECRET_ACCESS_KEY#aws-secret-access-key) keys. Please follow the documentation for access keys.

I share with you [an example usage](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-example-privatelink.html?highlight=endpoint_url) for endpoint_url option. [The changes](https://github.com/scrapy/scrapy/pull/4998/files) of PR #4998 pass a variable to [the function](https://github.com/boto/botocore/blob/develop/botocore/session.py#L727) via kwargs.

**Comment by lljr** [comment]
Thanks for the response @bkayranci ... It's still not clear how this all works together. I'm aware that the access key and the secret can be set in `settings.py`. I'm not sure how to set up `AWS_ENDPOINT_URL`. Should this be set in `settings.py` and how? 

The reason for my question is that I tried using Linode's Object Storage. I set `AWS_ENDPOINT_URL` to Linode's object storage URL. After I deploy, the file is not created...

Do I need to manually set boto with `AWS_ENDPOINT_URL`? I ask because the [function](https://github.com/boto/botocore/blob/develop/botocore/session.py#L727) linked in your response contains several parameters including the ones for access key and secret... 

How does it all tie together? I've never used boto and it's barely mentioned in the documentation. 

I'm asking how this all works together because it's confusing to set the access key and secret in a variable in `settings.py` but then seeing that there are parameters for these same ones in `create_client`.

**Comment by bkayranci** [comment]
> Thanks for the response @bkayranci ... It's still not clear how this all works together. I'm aware that the access key and the secret can be set in `settings.py`. I'm not sure how to set up `AWS_ENDPOINT_URL`. Should this be set in `settings.py` and how?
> 
> The reason for my question is that I tried using Linode's Object Storage. I set `AWS_ENDPOINT_URL` to Linode's object storage URL. After I deploy, the file is not created...
> 
> Do I need to manually set boto with `AWS_ENDPOINT_URL`? I ask because the [function](https://github.com/boto/botocore/blob/develop/botocore/session.py#L727) linked in your response contains several parameters including the ones for access key and secret...
> 
> How does it all tie together? I've never used boto and it's barely mentioned in the documentation.
> 
> I'm asking how this all works together because it's confusing to set the access key and secret in a variable in `settings.py` but then seeing that there are parameters for these same ones in `create_client`.

You can set in settings.py the AWS_ENDPOINT_URL variable.

You can find how to set the url in documentation of your cloud provider. For example, [DigitalOcean provides compatible api for s3.](https://docs.digitalocean.com/reference/api/spaces-api/#aws-s3-compatibility-1)

I set the `AWS_ENDPOINT_URL` variable in settings.py following,
```python
## settings.py
...
AWS_ENDPOINT_URL = 'https://nyc3.digitaloceanspaces.com'
...
```

Now, you are able to use it for storage backend. [docs](https://docs.scrapy.org/en/latest/topics/feed-exports.html?highlight=s3#s3)

**Comment by lljr** [comment]
@bkayranci Thanks for the response. My bad I deleted the old comment because I thought I had solved the issue. Anyway, I'm not able to post to the custom endpoint. I was able to trigger the action when I set the s3 uri in `FEEDS`. However, I got this error:  `botocore.exceptions.ClientError: An error occurred (InvalidAccessKeyId) when calling the PutObject operation: The AWS Access Key Id you provided does not exist in our records.`.

How do I trigger the post to my custom endpoint? Do I set the uri in `FEEDS` with the path to the file I want to store?

**Comment by bkayranci** [comment]
> @bkayranci Thanks for the response. My bad I deleted the old comment because I thought I had solved the issue. Anyway, I'm not able to post to the custom endpoint. I was able to trigger the action when I set the s3 uri in `FEEDS`. However, I got this error: `botocore.exceptions.ClientError: An error occurred (InvalidAccessKeyId) when calling the PutObject operation: The AWS Access Key Id you provided does not exist in our records.`.
> 
> How do I trigger the post to my custom endpoint? Do I set the uri in `FEEDS` with the path to the file I want to store?

I did not try the set endpoint in URI. My URI includes only bucket name and file path like "s3://mybucket/path/to/export.csv".

Also, do you sure that your environment includes the changes of #4998 ?

**Comment by lljr** [comment]
> Also, do you sure that your environment includes the changes of #4998 ?

Do you mean if I'm up to date? I'm using v2.5.0... 

> I did not try the set endpoint in URI. My URI includes only bucket name and file path like "s3://mybucket/path/to/export.csv".

Yes, that's what I mean. When I do this it triggers the post but it errors with an incorrect access key. I'll try to make a new bucket and use it's new keys instead. So far this is what I have and it works but it errors with the last error I just mentioned: 
```python
config = configparser.ConfigParser()
config_file = os.path.join(os.path.dirname(__file__), 'env.conf')
config.read(config_file)
AWS_ACCESS_KEY_ID = config['credentials']['AWS_ACCESS_KEY_ID']
AWS_SECRET_ACCESS_KEY = config['credentials']['AWS_SECRET_ACCESS_KEY']
AWS_ENDPOINT_URL = 'my-linode-bucket-test.us-southeast-1.linodeobjects.com'
AWS_USE_SSL = True
AWS_VERIFY = True

# HERE FOR DEBUGGIN PURPOSES
DEPTH_LIMIT = 1
FEEDS = {
    # 'epidemiology-data.json': {
    #     'format': 'json',
    #     'encoding': 'utf-8',
    #     'indent': 4,
    #     'overwrite': True,
    # },
    's3://epidemiology/data.json': {
        'format': 'json',
        'encoding': 'utf-8',
        'indent': 4,
        'overwrite': True,
    },
}
```

**Comment by wRAR** [comment]
> Do you mean if I'm up to date? I'm using v2.5.0...

Then it indeed doesn't.

**Comment by lljr** [comment]
@wRAR how do I download the latest? I used `python3 -m pip install git+https://github.com/scrapy/scrapy.git`?

**Comment by bkayranci** [comment]
> > Do you mean if I'm up to date? I'm using v2.5.0...
> 
> Then it indeed doesn't.


So, @lljr  sents to AWS with his secrets. @lljr  please, sure that remove your token, it could be leak.

@wRAR Is there anything you expect from me for including in new version?

**Comment by lljr** [comment]
@bkayranci I didn't ... I read them from a local file

**Comment by bkayranci** [comment]
> @bkayranci I didn't ... I read them from a local file

`botocore.exceptions.ClientError: An error occurred (InvalidAccessKeyId) when calling the PutObject operation: The AWS Access Key Id you provided does not exist in our records.`

it means you could not pass to the function your endpoint. so, your requests sent to AWS. Now, AWS could be know your token.

**Comment by lljr** [comment]
@bkayranci yes, ok so maybe this is just that Linode is not compatible? I read [here](https://www.linode.com/docs/products/storage/object-storage/guides/aws-sdk-for-python/)

`[cluster-id]` needs to be set therefore my `end_point` url is wrong. I was able to fix this and in the stacktrace I can see 

```
2021-08-30 12:53:45 [botocore.auth] DEBUG: CanonicalRequest:
PUT
/epidemiology-ni/epidemiology-data-ni.json
```

but it errors with `botocore.exceptions.ClientError: An error occurred (InvalidAccessKeyId) when calling the PutObject operation: Unknown`. Maybe this is not compatible with Linode? Let me know your thoughts... BTW the access key is set (given by Linode).

**Comment by wRAR** [comment]
> how do I download the latest? I used `python3 -m pip install git+https://github.com/scrapy/scrapy.git`?

That should work.

> Is there anything you expect from me for including in new version?

No, it was merged and so will be included in the new version.

**Comment by lljr** [comment]
It works. I had to write in the credentials in the s3 uri like in the [documentation](https://docs.scrapy.org/en/latest/topics/feed-exports.html#s3) like this `s3://aws_key:aws_secret@mybucket/path/to/export.csv`. 

This won't work
```python
AWS_ACCESS_KEY_ID = config['credentials']['AWS_ACCESS_KEY_ID']
AWS_SECRET_ACCESS_KEY = config['credentials']['AWS_SECRET_ACCESS_KEY']
``` 

Any idea why? Or at least it could help telling me how to debug?? I use to have the aws cli installed in my computer. I deleted the `~/.aws` folder since it's no longer in use. I think boto looks into this folder first... I'm not sure but even deleting this folder would throw the same invalid secret key error.

**Pull request merged**